### PR TITLE
[1638] Align the lifecycle of the views on the lifecycle of the domains

### LINKED ISSUE #1638: Align the lifecycle of the views on the lifecycle of the domains

### PR DESCRIPTION
**Pull request by sbegaudeau**
Bug: https://github.com/eclipse-sirius/sirius-components/issues/1638
Signed-off-by: Stéphane Bégaudeau <stephane.begaudeau@obeo.fr>

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/core/backend/sirius-components-collaborative/src/main/java/org/eclipse/sirius/components/collaborative/editingcontext/EditingContextEventProcessor.java [hunk 1]
```diff
@@ -326,6 +321,11 @@ private void handleRepresentationInput(One<IPayload> payloadSink, IRepresentatio
         } else {
             this.logger.warn("No representation event processor found for event: {}", representationInput);
         }
+
+        this.editingContextEventHandlers.stream()
+                .filter(handler -> handler.canHandle(this.editingContext, representationInput))
+                .findFirst()
+                .ifPresent(handler -> handler.handle(payloadSink, this.changeDescriptionSink, this.editingContext, representationInput));
```
Comment (AxelRICHARD) [inline packages/core/backend/sirius-components-collaborative/src/main/java/org/eclipse/sirius/components/collaborative/editingcontext/EditingContextEventProcessor.java]: Why do you need to do this in `handleRepresentationInput`?
What if the handler is not present? (it is handled in `handleInput` but not here)

File: packages/sirius-web/backend/sirius-web-services/src/main/java/org/eclipse/sirius/web/services/editingcontext/RepresentationDescriptionSearchService.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * Service used to search the representation descriptions available.
+ *
+ * @author sbegaudeau
+ * @author hmarchadour
+ */
+@Service
+public class RepresentationDescriptionSearchService implements IRepresentationDescriptionSearchService {
+
+    @Override
+    public Map<String, IRepresentationDescription> findAll(IEditingContext editingContext) {
```
Comment (AxelRICHARD) [inline packages/sirius-web/backend/sirius-web-services/src/main/java/org/eclipse/sirius/web/services/editingcontext/RepresentationDescriptionSearchService.java:34]: Question: Is there a case where `IEditingContext ` does not need to know  the representation descriptions?

It seems that every implementation of `IRepresentationDescriptionSearchService` will need to have access to a map of `IRepresentationDescription `, so why `IEditingContext` does not expose such method?

File: packages/sirius-web/backend/sirius-web-services/src/main/java/org/eclipse/sirius/web/services/editingcontext/EditingContextSearchService.java [hunk 3]
```diff
@@ -107,10 +119,17 @@ public Optional<IEditingContext> findById(String editingContextId) {
...
+        this.configurers.forEach(configurer -> configurer.addRepresentationDescriptions(registry));
+        registry.getRepresentationDescriptions().forEach(representationDescription -> representationDescriptions.put(representationDescription.getId(), representationDescription));
+        this.dynamicRepresentationDescriptionService.findDynamicRepresentationDescriptions(editingContextId, editingDomain)
+                .forEach(representationDescription -> representationDescriptions.put(representationDescription.getId(), representationDescription));
+
         long end = System.currentTimeMillis();
         this.timer.record(end - start, TimeUnit.MILLISECONDS);
 
-        return Optional.of(new EditingContext(editingContextId, editingDomain));
+        return Optional.of(new EditingContext(editingContextId, editingDomain, representationDescriptions));
```
Comment (AxelRICHARD) [inline packages/sirius-web/backend/sirius-web-services/src/main/java/org/eclipse/sirius/web/services/editingcontext/EditingContextSearchService.java:132]: Is the `IEditingContext` used by some query/mutation/subscription without the need of representation descriptions? If it is the case, may be the computation of existing representation descriptions could be computed in  implementations of `IRepresentationDescriptionSearchService` ?

File: packages/core/backend/sirius-components-collaborative/src/main/java/org/eclipse/sirius/components/collaborative/dto/GetRepresentationDescriptionInput.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.UUID;
+
+import org.eclipse.sirius.components.core.api.IRepresentationInput;
+
+/**
+ * Used to retrieve a representation description.
+ *
+ * @author sbegaudeau
+ */
+public record GetRepresentationDescriptionInput(UUID id, String editingContextId, String representationId) implements IRepresentationInput {
```
Comment (sbegaudeau) [inline packages/core/backend/sirius-components-collaborative/src/main/java/org/eclipse/sirius/components/collaborative/dto/GetRepresentationDescriptionInput.java]: Should be an `IEditingContextInput`

File: packages/formdescriptioneditors/backend/sirius-components-formdescriptioneditors-graphql/src/main/java/org/eclipse/sirius/components/formdescriptioneditors/graphql/datafetchers/subscription/SubscriptionFormDescriptionEditorEventDataFetcher.java [hunk 5]
```diff
@@ -43,31 +46,37 @@
...
-        // @formatter:on
+        Map<String, Object> localContext = new HashMap<>();
+        localContext.put(LocalContextConstants.EDITING_CONTEXT_ID, input.editingContextId());
+        localContext.put(LocalContextConstants.REPRESENTATION_ID, input.formDescriptionEditorId());
+
+        return this.exceptionWrapper.wrapFlux(() -> this.eventProcessorSubscriptionProvider.getSubscription(input.editingContextId(), IFormDescriptionEditorEventProcessor.class, formDescriptionEditorConfiguration, input), input)
+                .map(payload ->  DataFetcherResult.<IPayload>newResult()
+                        .data(payload)
+                        .localContext(localContext)
+                        .build());
```
Comment (AxelRICHARD) [inline packages/formdescriptioneditors/backend/sirius-components-formdescriptioneditors-graphql/src/main/java/org/eclipse/sirius/components/formdescriptioneditors/graphql/datafetchers/subscription/SubscriptionFormDescriptionEditorEventDataFetcher.java:79]: Those lines seem not aligned with others.

File: packages/selection/backend/sirius-components-selection-graphql/src/main/java/org/eclipse/sirius/components/selection/graphql/datafetchers/subscription/SubscriptionSelectionEventDataFetcher.java [hunk 6]
```diff
@@ -52,14 +55,20 @@ public SubscriptionSelectionEventDataFetcher(ObjectMapper objectMapper, IExcepti
...
-                input);
+        Map<String, Object> localContext = new HashMap<>();
+        localContext.put(LocalContextConstants.EDITING_CONTEXT_ID, input.editingContextId());
+        localContext.put(LocalContextConstants.REPRESENTATION_ID, selectionConfiguration.getId());
 
+        return this.exceptionWrapper.wrapFlux(() -> this.eventProcessorSubscriptionProvider.getSubscription(input.editingContextId(), ISelectionEventProcessor.class, selectionConfiguration, input), input)
+                .map(payload ->  DataFetcherResult.<IPayload>newResult()
+                        .data(payload)
+                        .localContext(localContext)
+                        .build());
```
Comment (AxelRICHARD) [inline packages/selection/backend/sirius-components-selection-graphql/src/main/java/org/eclipse/sirius/components/selection/graphql/datafetchers/subscription/SubscriptionSelectionEventDataFetcher.java:71]: Those lines seem not aligned with others.

File: packages/diagrams/backend/sirius-components-diagrams-graphql/src/main/java/org/eclipse/sirius/components/diagrams/graphql/datafetchers/subscription/SubscriptionDiagramEventDataFetcher.java [hunk 7]
```diff
@@ -52,12 +55,19 @@ public SubscriptionDiagramEventDataFetcher(ObjectMapper objectMapper, IException
...
-                input);
+        Map<String, Object> localContext = new HashMap<>();
+        localContext.put(LocalContextConstants.EDITING_CONTEXT_ID, input.editingContextId());
+        localContext.put(LocalContextConstants.REPRESENTATION_ID, input.diagramId());
+
+        return this.exceptionWrapper.wrapFlux(() -> this.eventProcessorSubscriptionProvider.getSubscription(input.editingContextId(), IDiagramEventProcessor.class, diagramConfiguration, input), input)
+                .map(payload ->  DataFetcherResult.<IPayload>newResult()
+                        .data(payload)
+                        .localContext(localContext)
+                        .build());
```
Comment (AxelRICHARD) [inline packages/diagrams/backend/sirius-components-diagrams-graphql/src/main/java/org/eclipse/sirius/components/diagrams/graphql/datafetchers/subscription/SubscriptionDiagramEventDataFetcher.java:71]: Those lines seem not aligned with others.

File: packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionFormEventDataFetcher.java [hunk 8]
```diff
@@ -52,12 +55,19 @@ public SubscriptionFormEventDataFetcher(ObjectMapper objectMapper, IExceptionWra
...
-    }
+        Map<String, Object> localContext = new HashMap<>();
+        localContext.put(LocalContextConstants.EDITING_CONTEXT_ID, input.editingContextId());
+        localContext.put(LocalContextConstants.REPRESENTATION_ID, input.formId());
 
+        return this.exceptionWrapper.wrapFlux(() -> this.eventProcessorSubscriptionProvider.getSubscription(input.editingContextId(), IFormEventProcessor.class, formConfiguration, input), input)
+                .map(payload ->  DataFetcherResult.<IPayload>newResult()
+                        .data(payload)
+                        .localContext(localContext)
+                        .build());
```
Comment (AxelRICHARD) [inline packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionFormEventDataFetcher.java:71]: Those lines seem not aligned with others.

File: packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionPropertiesEventDataFetcher.java [hunk 9]
```diff
@@ -51,13 +54,20 @@ public SubscriptionPropertiesEventDataFetcher(ObjectMapper objectMapper, IExcept
...
-                input);
+        Map<String, Object> localContext = new HashMap<>();
+        localContext.put(LocalContextConstants.EDITING_CONTEXT_ID, input.editingContextId());
+        localContext.put(LocalContextConstants.REPRESENTATION_ID, propertiesConfiguration.getId());
+
+        return this.exceptionWrapper.wrapFlux(() -> this.eventProcessorSubscriptionProvider.getSubscription(input.editingContextId(), IFormEventProcessor.class, propertiesConfiguration, input), input)
+                .map(payload ->  DataFetcherResult.<IPayload>newResult()
+                        .data(payload)
+                        .localContext(localContext)
+                        .build());
```
Comment (AxelRICHARD) [inline packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionPropertiesEventDataFetcher.java:70]: Those lines seem not aligned with others.

File: packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionRelatedElementsEventDataFetcher.java [hunk 10]
```diff
@@ -51,13 +54,20 @@ public SubscriptionRelatedElementsEventDataFetcher(ObjectMapper objectMapper, IE
...
-                input);
+        Map<String, Object> localContext = new HashMap<>();
+        localContext.put(LocalContextConstants.EDITING_CONTEXT_ID, input.editingContextId());
+        localContext.put(LocalContextConstants.REPRESENTATION_ID, propertiesConfiguration.getId());
+
+        return this.exceptionWrapper.wrapFlux(() -> this.eventProcessorSubscriptionProvider.getSubscription(input.editingContextId(), IFormEventProcessor.class, propertiesConfiguration, input), input)
+                .map(payload ->  DataFetcherResult.<IPayload>newResult()
+                        .data(payload)
+                        .localContext(localContext)
+                        .build());
```
Comment (AxelRICHARD) [inline packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionRelatedElementsEventDataFetcher.java:70]: Those lines seem not aligned with others.

File: packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionRepresentationsEventDataFetcher.java [hunk 11]
```diff
@@ -51,13 +54,20 @@ public SubscriptionRepresentationsEventDataFetcher(ObjectMapper objectMapper, IE
...
-                input);
+        Map<String, Object> localContext = new HashMap<>();
+        localContext.put(LocalContextConstants.EDITING_CONTEXT_ID, input.editingContextId());
+        localContext.put(LocalContextConstants.REPRESENTATION_ID, representationsConfiguration.getId());
+
+        return this.exceptionWrapper.wrapFlux(() -> this.eventProcessorSubscriptionProvider.getSubscription(input.editingContextId(), IFormEventProcessor.class, representationsConfiguration, input), input)
+                .map(payload ->  DataFetcherResult.<IPayload>newResult()
+                        .data(payload)
+                        .localContext(localContext)
+                        .build());
```
Comment (AxelRICHARD) [inline packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionRepresentationsEventDataFetcher.java:70]: Those lines seem not aligned with others.

File: packages/sirius-web/backend/sirius-web-graphql/src/main/java/org/eclipse/sirius/web/graphql/datafetchers/subscriptions/SubscriptionEditingContextEventDataFetcher.java [hunk 12]
```diff
@@ -53,15 +57,20 @@ public SubscriptionEditingContextEventDataFetcher(ObjectMapper objectMapper, IEd
...
+
         return this.editingContextEventProcessorRegistry.getOrCreateEditingContextEventProcessor(input.editingContextId())
                 .map(IEditingContextEventProcessor::getOutputEvents)
-                .orElse(Flux.empty());
-        // @formatter:on
+                .orElse(Flux.empty())
+                .map(payload ->  DataFetcherResult.<IPayload>newResult()
+                        .data(payload)
+                        .localContext(localContext)
+                        .build());
```
Comment (AxelRICHARD) [inline packages/sirius-web/backend/sirius-web-graphql/src/main/java/org/eclipse/sirius/web/graphql/datafetchers/subscriptions/SubscriptionEditingContextEventDataFetcher.java:73]: Those lines seem not aligned with others.

File: packages/formdescriptioneditors/backend/sirius-components-formdescriptioneditors-graphql/src/main/java/org/eclipse/sirius/components/formdescriptioneditors/graphql/datafetchers/subscription/SubscriptionFormDescriptionEditorEventDataFetcher.java (hunk 5 above)
Comment (sbegaudeau) [reply packages/formdescriptioneditors/backend/sirius-components-formdescriptioneditors-graphql/src/main/java/org/eclipse/sirius/components/formdescriptioneditors/graphql/datafetchers/subscription/SubscriptionFormDescriptionEditorEventDataFetcher.java:79]: They are properly aligned

File: packages/selection/backend/sirius-components-selection-graphql/src/main/java/org/eclipse/sirius/components/selection/graphql/datafetchers/subscription/SubscriptionSelectionEventDataFetcher.java (hunk 6 above)
Comment (sbegaudeau) [reply packages/selection/backend/sirius-components-selection-graphql/src/main/java/org/eclipse/sirius/components/selection/graphql/datafetchers/subscription/SubscriptionSelectionEventDataFetcher.java:71]: Same comment

File: packages/diagrams/backend/sirius-components-diagrams-graphql/src/main/java/org/eclipse/sirius/components/diagrams/graphql/datafetchers/subscription/SubscriptionDiagramEventDataFetcher.java (hunk 7 above)
Comment (sbegaudeau) [reply packages/diagrams/backend/sirius-components-diagrams-graphql/src/main/java/org/eclipse/sirius/components/diagrams/graphql/datafetchers/subscription/SubscriptionDiagramEventDataFetcher.java:71]: Same comment

File: packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionFormEventDataFetcher.java (hunk 8 above)
Comment (sbegaudeau) [reply packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionFormEventDataFetcher.java:71]: Same comment

File: packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionRelatedElementsEventDataFetcher.java (hunk 10 above)
Comment (sbegaudeau) [reply packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionRelatedElementsEventDataFetcher.java:70]: Same comment

File: packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionPropertiesEventDataFetcher.java (hunk 9 above)
Comment (sbegaudeau) [reply packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionPropertiesEventDataFetcher.java:70]: Same comment

File: packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionRepresentationsEventDataFetcher.java (hunk 11 above)
Comment (sbegaudeau) [reply packages/forms/backend/sirius-components-forms-graphql/src/main/java/org/eclipse/sirius/components/forms/graphql/datafetchers/subscription/SubscriptionRepresentationsEventDataFetcher.java:70]: Same comment

File: packages/sirius-web/backend/sirius-web-graphql/src/main/java/org/eclipse/sirius/web/graphql/datafetchers/subscriptions/SubscriptionEditingContextEventDataFetcher.java (hunk 12 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-graphql/src/main/java/org/eclipse/sirius/web/graphql/datafetchers/subscriptions/SubscriptionEditingContextEventDataFetcher.java:73]: Same comment

**Comment by sbegaudeau** [comment]
There is an useless deadlock in the form description editor, I'll fix that and update the PR

**Review by AxelRICHARD — APPROVED** [review]

**Pull request merged**